boards/p1am_100/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+    PA24 {
+        name: usb_dm,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDm
+        }
+    }
+    PA25 {
+        name: usb_dp,
+        aliases: {
+            #[cfg(feature = "usb")]
+            AlternateG: UsbDp
+        }
+    }
+    PA27 {
+        name: pa27,
+        aliases: {
+            PullUpInput: SdCardDetect
+        }
+    }
+    PA28 {
+        name: switch,
+        aliases: {
+            PullUpInput: Switch
+        }
+    }
+    PB02 {
+        name: d16,
+        aliases: {
+            AlternateB: A1
+        }
+    }
+    PB03 {
+        name: d17,
+        aliases: {
+            AlternateB: A2
+        }
+    }
+    PB08 {
+        name: led,
+        aliases: {
+            PushPullOutput: Led
+        }
+    }
+    PB09 {
+        name: pb09,
+        aliases: {
+            PushPullOutput: BaseEnable,
+            AlternateB: AdcBattery
+        }
+    }
+    PB10 { name: d4 }
+    PB11 { name: d5 }
+    PB22 {
+        name: d14,
+        aliases: {
+            AlternateD: UartTx
+        }
+    }
+    PB23 {
+        name: d13,
+        aliases: {
+            AlternateD: UartRx
+        }
+    }
+);
+
+const BASE_CONTROLLER_FREQ: Hertz = Hertz(1000000);
+// FIXME: const BASE_CONTROLLER_SPI_MODE: dyn spi::Mode = &spi::MODE_2;
+
+/// Convenience for setting up the labeled SPI0 peripheral.
+/// SPI0 has the P1AM base controller connected.
+/// This powers up SERCOM1 and configures it for talking to the
+/// base controller.
+pub fn base_controller_spi(
+    clocks: &mut GenericClockController,
+    sercom1: pac::SERCOM1,
+    pm: &mut pac::PM,
+    sck: Spi0Sck,
+    mosi: Spi0Mosi,
+    miso: Spi0Miso,
+) -> sercom::v2::spi::Spi<impl spi::ValidConfig> {
+    let gclk0 = &clocks.gclk0();
+    let core_clock = &clocks.sercom1_core(&gclk0).unwrap();
+    let pads = spi::Pads::new().sclk(sck).data_in(miso).data_out(mosi);
+    spi::Config::new(pm, sercom1, pads, core_clock.freq())
+        .baud(BASE_CONTROLLER_FREQ)
+        .spi_mode(spi::MODE_2)
+        .enable()
+}
+
+/// Convenience for setting up the labelled SDA, SCL pins to
+/// operate as an I2C master running at the specified frequency.
+pub fn i2c_master<F: Into<Hertz>>(
+    clocks: &mut GenericClockController,
+    bus_speed: F,
+    sercom0: pac::SERCOM0,
+    pm: &mut pac::PM,
+    sda: Sda,
+    scl: Scl,
+    port: &mut Port,
+) -> I2CMaster0<
+    impl hal::sercom::AnyPad<Sercom = pac::SERCOM0>,
+    impl hal::sercom::AnyPad<Sercom = pac::SERCOM0>,
+> {
+    let gclk0 = clocks.gclk0();
+    I2CMaster0::new(
+        &clocks.sercom0_core(&gclk0).unwrap(),
+        bus_speed.into(),
+        sercom0,
+        pm,
+        <hal::gpio::v1::Pin<_, _>>::from(sda).into_pad(port),
+        <hal::gpio::v1::Pin<_, _>>::from(scl).into_pad(port),
+    )
+}
+
+pub type Uart = UART5<
+        hal::sercom::Sercom5Pad3<<UartRx as AnyPin>::Id>,
+    hal::sercom::Sercom5Pad2<<UartTx as AnyPin>::Id>,
```

Comment:
FYI, this only works right now because the `v1` and `v2` pad types are mixed. The `Sercom5Pad3`, etc. aliases are for the `v1` pad types. They take a complete pin in their type parameter, e.g. `Sercom5Pad3<UartRx>`. The `v2` pad types only need the `PinId`. They look like this: `Pad<Sercom5, Pad3, UartRxPinId>`.

In the future update, your current code won't work. But it should be an easy fix.

Once the `v1` pad types are removed, maybe I'll recast the `SercomXPadY` type aliases to be `v2` pads. I could also consider adding a macro to automatically create `Pad` aliases, just like `Pin` aliases with `bsp_pins!`.

Reply:
Again I don't want to have to specify these pad numbers here at all, but I couldn't figure out how to avoid it since UART5 wants a `v1::Pad`.